icx/src/main.rs

```diff
@@ -139,49 +140,71 @@ struct PrincipalConvertOpts {
+        format!(
+            "Failed when check the candid file: {}",
+            idl_path.to_string_lossy()
+        )
+    })?;
+    match ty {
+        None => Ok(None),
+        Some(actor) => {
+            let method = env
+                .get_method(&actor, method_name)
+                .with_context(|| format!("Failed to get method: {}", method_name))?
+                .clone();
+            Ok(Some((env, method)))
+        }
+    }
 }
 
-pub fn check_candid_file(idl_path: &std::path::Path) -> Result<(TypeEnv, Option<Type>), String> {
-    let idl_file = std::fs::read_to_string(idl_path).map_err(|e| format!("{:?}", e))?;
-    let ast = idl_file
-        .parse::<IDLProg>()
-        .map_err(|e| format!("{:?}", e))?;
+pub fn check_candid_file(idl_path: &std::path::Path) -> Result<(TypeEnv, Option<Type>)> {
+    let idl_file = std::fs::read_to_string(idl_path)
+        .with_context(|| format!("Failed to read Candid file: {}", idl_path.to_string_lossy()))?;
+    let ast = idl_file.parse::<IDLProg>().with_context(|| {
+        format!(
+            "Failed to parse the Candid file: {}",
+            idl_path.to_string_lossy()
+        )
+    })?;
     let mut env = TypeEnv::new();
-    let actor = check_prog(&mut env, &ast).map_err(|e| format!("{:?}", e))?;
+    let actor = check_prog(&mut env, &ast).with_context(|| {
+        format!(
+            "Failed to type check the Candid file: {}",
+            idl_path.to_string_lossy()
+        )
+    })?;
     Ok((env, actor))
 }
 
 fn blob_from_arguments(
     arguments: Option<&str>,
     arg_type: &ArgType,
     method_type: &Option<(candid::parser::typing::TypeEnv, candid::types::Function)>,
-) -> Result<Vec<u8>, String> {
+) -> Result<Vec<u8>> {
     let mut buffer = Vec::new();
-    let args = if arguments == Some("-") {
+    let arguments = if arguments == Some("-") {
         use std::io::Read;
         std::io::stdin().read_to_end(&mut buffer).unwrap();
         std::str::from_utf8(&buffer).ok()
     } else {
         arguments
     };
+
     match arg_type {
         ArgType::Raw => {
-            let bytes = hex::decode(&args.unwrap_or(""))
-                .map_err(|e| format!("Argument is not a valid hex string: {}", e))?;
+            let bytes = hex::decode(&arguments.unwrap_or(""))
+                .context("Argument is not a valid hex string")?;
             Ok(bytes)
         }
         ArgType::Idl => {
-            let arguments = args.unwrap_or("()");
-            let args: Result<IDLArgs, String> = arguments
-                .parse::<IDLArgs>()
-                .map_err(|e: candid::Error| format!("Invalid Candid values: {}", e));
+            let arguments = arguments.unwrap_or("()");
+            let args = arguments.parse::<IDLArgs>();
             let typed_args = match method_type {
-                None => args?.to_bytes(),
+                None => args
+                    .context("Failed to parse arguments as Candid values")?
```

Comment:
Can we have this context message and the one below use different text, so that we can differentiate the cases?

Reply:
Sure, will do